### DISCUSSION TITLE
Sharing model for multiple workspaces

### DISCUSSION THREAD

**Discussion by hubertmis** [likec4/likec4#962] [Ideas]
I'm working on a few projects sharing the same source code, but in different configurations. I would like to create diagrams for each of these projects in a separate workspace.

Currently, I can create workspaces in different folders and create a symbolic link from each of the workspaces to another folder containing the model. However, I'm missing a feature to filter out elements and relations that do not apply to a project.

In structurizr I'm filtering based on tags:
```
exclude "element.tag==project2 && element.tag!=project1"
exclude "element.tag==project3 && element.tag!=project1"
```

and I `!include` this snippet to all my views in the project1 workspace.

I wish I had similar filtering infrastructure available in likec4. I can imagine a syntax like:
```
views {
  view1 {
    ...
  }

  view2 {
    ...
  }

  exclude "element.tag == #project2 && element.tag != #project1"
  exclude "element.tag == #project3 && element.tag != #project1"
}
```

with the additional `exclude` statements at the top `views` level applying to all the views in the workspace.

**Comment by mtaylor** [comment]
I am working on describing a project with a similar structure.  We have several shared services that provide common capabilities across our portfolio.  We would like to describe the architecture of those services once, and include them in a number of systems.  (Note in Structurizr we work around this by using the include directive (and create snippets that describes each service) 

If there was an ability to provide a reference to an element within an element structure that would be perfect.  For example: 

```
    commonSharedService = container "Common capability shared across services"
    
    System1 = system {
        include commonSharedService
    }

    System2 = system {
        include commonSharedService
    }
```

**Reply by hubertmis** [reply]
I'm working on a solution to achieve something similar, but with a little different approach.

I plan to make the following folder structure:

```
likec4-files/
|
|-+ common/
| \- multiple folders and files describing the model, potentially spec and views
|
\-+ projects/
  |
  |-+ project1/
  | |- link to ../../common
  | \- workspace.likec4
  |
  \-+ project2/
    |- link to ../../common
    \- workspace.likec4
```

The content of both `workspace.likec4` would be something like (swapping project ids of course):
```
global {
  incExcGroup project_predicates {
    exclude * where tag is #project2 and tag is not #project1 // Remove all elements tagged to be used in project2, but not tagged to be used in project1
  }
}
```

And the elements in the model would contain tagging:
```
model {
  component common_component
  component uart_driver {
    #project1
  }
  component spi_driver {
    #project2
  }

  common_component -> uart_driver 'send data' #project1
  common_component -> spi_driver 'send data' #project2
}
```

And all the `views` blocks would include:
```
views {
  global incExc project_predicates

  view { ... }
}
```

I'm not sure yet if the specification should be in the common or project-specific folders.
I'll define common views in the common folder and the project-specific views in the project specific folder.

Would a solution like this work also for you?

**Reply by mtaylor** [reply]
Thanks for this @hubertmis.  

Using symlinks is a good idea, that might solve part of the problem I am facing.  This could work in combination with #533 (depending on the solution).

**Reply by rumpelrausch** [reply]
In the context I'm working with symlinks are not a preferred solution.

Our workflow starts with fast development using VSCode, implements code reviews and finally pipeline-based deployments as static websites.
Build/deployment pipelines and some web servers might choke on symlinks; In addition symlinks are handled by a filesystem layer, not natively in LikeC4 / JS / TS. De-referencing them in node JS is possible, but you'd definitely leave the domain of "document something".

In other words:
Symlinks are a technical approach when instead we should talk about functional requirements.

Currently we pull global definitions from an npm package (not public) which copies a "shared-globals" folder into the local LikeC4 workspace. It works, but an extension of the DSL allowing relative file includes would be more appropriate.
Another pro issue: Relocation of workspace folders would be supported by the VSCode IDE if the DSL understood external file dependencies.

Using it that way would keep all knowledge about location of files _inside_ the  documents.
Everything would burn down to scaffolding rules, and you could still define them on your own.

**Reply by hubertmis** [reply]
> symlinks are handled by a filesystem layer, not natively in LikeC4 / JS / TS

I was surprised by this, but it turned out the symlinks are not handled natively. They are handled by a filesystem handler in JS / TS. The default filesystem handler from Langium does not follow symlinks so I've [overridden it](https://github.com/likec4/likec4/pull/1213) to follow my plan. So it looks like the symlinks could be a solution to meet the requirement to share parts of models. It is extensible to other environments.

However, I agree it's not an ultimate solution and another one can be introduced to realize scenarios in which symlinks don't meet the requirements.

**Comment by davydkov** [comment]
@hubertmis 
Please check https://github.com/likec4/likec4/discussions/1674
